### PR TITLE
Added support for response wrapping and unwrapping.

### PR DESCRIPTION
**Pull request by singram**
Added rudimentary support for response wrapping and unwrapping.

Main focus was to facilitate secure introductions.

Probably needs a clearer way to define the ttl at the user API.  Considered putting it in 'options' but thought I'd solicit feedback first.  The auth_token creation methods could also be dried up depending on final approach.

Fixed minor problem with overriding the token header in client.rb.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -39,3 +39,5 @@ Gemfile.lock
 
 # Project-specific
 spec/tmp
+/vault
```
Comment (sethvargo) [inline .gitignore]: Please remove these lines. Thanks!

File: lib/vault/api/logical.rb [hunk 2]
```diff
@@ -40,7 +40,7 @@ def list(path)
     # @param [String] path
     #   the path to read
     #
-    # @return [Secret, nil]
+    # @return [Vault::Secret, nil]
```
Comment (sethvargo) [inline lib/vault/api/logical.rb]: Please do not introduce unnecessary changes like this. Thanks!

File: lib/vault/api/logical.rb [hunk 3]
```diff
@@ -84,5 +84,43 @@ def delete(path)
...
+    #
+    # @return [String]
+    def unwrap_token(wrapper_token)
+      wrapper_token = wrapper_token.wrap_info.token if wrapper_token.is_a?(Vault::Secret) && wrapper_token.wrap_info
+      unwrapped_token_response = unwrap(wrapper_token)
+      return unwrapped_token_response.data.auth.client_token
+    rescue HTTPError => e
+      raise
+    end
+
```
Comment (sethvargo) [inline lib/vault/api/logical.rb]: Extra newline

File: lib/vault/client.rb [hunk 4]
```diff
@@ -16,6 +16,9 @@ class Client
     # The name of the header used to hold the Vault token.
     TOKEN_HEADER = "X-Vault-Token".freeze
 
+    # The name of the header used to hold the wrapped request ttl.
+    WRAP_TTL_HEADER = "X-Vault-Wrap-Ttl".freeze
```
Comment (sethvargo) [inline lib/vault/client.rb]: I think this is actually `X-Vault-Wrap-TTL` (note all caps on TTL)

File: lib/vault/client.rb [hunk 5]
```diff
@@ -132,7 +135,7 @@ def request(verb, path, data = {}, headers = {})
       # Add the Vault token header - users could still override this on a
       # per-request basis
       if !token.nil?
-        request.add_field(TOKEN_HEADER, token)
+        headers[TOKEN_HEADER] ||= token
```
Comment (sethvargo) [inline lib/vault/client.rb]: Why change this?

**Comment by sethvargo** [comment]
Hi there,

Thank you for the pull request. I really think `wrap_ttl` should just be an option to the existing options hash instead of introducing a secondary option to the method. This would make the signature:

```ruby
Vault.token.create(wrap_ttl: 50)
```

which seems like a nicer UX to me. What do you think?

There are also some unnecessary changes to the gitignore and some YARD comments. Please revert those. Thanks!

File: lib/vault/client.rb (hunk 5 above)
Comment (singram) [inline lib/vault/client.rb]: Experiments with overriding this header field via the header parameters hash were not working as expected (ruby 2.3.0).  This change fixed that allowing the user to override the value, as per the expectation, as well as consolidating the generation of headers on the request object.

File: lib/vault/api/logical.rb (hunk 2 above)
Comment (singram) [inline lib/vault/api/logical.rb]: Np.  I didn't see a documentation tool in the gemspec but was trying to align the expectations with the other documentation which fully qualify the Secret object.

**Comment by singram** [comment]
Yeah I agree it provides a clearer UX experience for the user.  Any reason not to add yard as part of the developer gemspec so that documentation can be linted etc?

File: lib/vault/client.rb (hunk 4 above)
Comment (singram) [inline lib/vault/client.rb]: Good catch.  Looks like the http proxy I was using for equivalence testing with the go client did some case manipulation.

**Comment by singram** [comment]
@sethvargo All comments should be addressed. Yard was also added as a development dependency and one resulting documentation error fixed.  Let me know if you see anything else that needs addressed.
